## Supervised Learning

### This is a type of machine learning that uses algorithm, where the algorithm learns from labeled trained data to make predictions or decisions.

### LABELED Data in supervised learning is like having answers (labels) provided along with the data, so the computer can learn how to make predictions. UNLABELED Data in unsupervised learning is about discovering patterns and relationships within the data without any initial labels, helping to uncover hidden insights.

### 

# Type of models i used :

## Neural Network : 
###               is a computer system inspired by how our brain works. It's designed to learn from examples and make decisions, just like we do.so basical it functions the way a human brain does.

## Decision Tree model:
###                a decision tree helps a computer make decisions by asking a series of questions about the input data. Each question leads to more questions or a final answer. It's a way for a computer to "learn" from examples and figure out the best path to make predictions or classify things.It basical deals classifications and regression.

## Linear Regresssion:
###               is used to model the relationship between a dependent variable and one or more independent variables by fitting a linear equation

# Neural  Networks 

## Why i used Neural Networks

### At first i didnt really have a reason but on doing some research i figured out that Neural network is used best in image classification i.e neural networks in supervised learning basically makes computers to become smart at recognizing things, like animals, by learning from labeled examples.So in this situaton it is not really ideal.

### During my observations i noticed that when considering Job Title and Education Level as factors and using the mean score (Also known as R^2 Score, measures the score ranges from 0 to 1, where a higher value shows that its performing well in predicting), the predictions often fall below zero. In other words, the predicted values are relatively low, as they tend to be closer to zero than to one.

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPRegressor
from sklearn.preprocessing import OneHotEncoder, LabelEncoder
from sklearn.tree import DecisionTreeRegressor, DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LinearRegression
from sklearn.compose import ColumnTransformer
from sklearn.metrics import mean_absolute_error

###  Using the databse given the prediction below is Salary based on Age 

In [2]:
# Reading the data from the database
salary_df = pd.read_csv("Clean Database.csv")

# features and target variables
X_age_to_salary = salary_df[['Age']]
y_salary = salary_df['Salary']

# Spliting the data into training and testing sets
X_train_age_to_salary, X_test_age_to_salary, y_train_salary, y_test_salary = train_test_split(X_age_to_salary, y_salary, test_size=0.2)

# Initializing the MLPRegressor models
model_age_salary = MLPRegressor(hidden_layer_sizes=(50, 50), max_iter=1000)

# training the MLPRegressor models
model_age_salary.fit(X_train_age_to_salary, y_train_salary)

# Make predictions
predictions_salary = model_age_salary.predict(X_test_age_to_salary)

# Print the predictions
# print("Predicted Salary based on Age:", predictions_salary)

# Putting predictions in a tabular format
predictions_df = pd.DataFrame({
    'Predicted Salary based on Age': predictions_salary
})
print(predictions_df)
print("Mean Score Predictions: ", model_age_salary.score(X_test_age_to_salary, y_test_salary))


     Predicted Salary based on Age
0                    145386.306207
1                    115225.477232
2                    142035.102987
3                    135332.696549
4                    135332.696549
..                             ...
353                   85064.648257
354                   91767.054696
355                   91767.054696
356                   98469.461134
357                   98469.461134

[358 rows x 1 columns]
Mean Score Predictions:  0.5603820520229996


C:\Users\HP\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:684: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(


### Predictions of Years of Experience based on Age

In [3]:
# Reading the data
salary_df = pd.read_csv("Clean Database.csv")

# features and target variables
X_age_to_yearsofexp = salary_df[['Age']]
y_year_of_experience = salary_df[['Years of Experience']]


# Split the data into training and testing sets
X_train_age_to_yearsofexp, X_test_age_to_yearsofexp, y_train_year_of_experience, y_test_year_of_experience = train_test_split(X_age_to_yearsofexp, y_year_of_experience, test_size=0.2)


# Initializing the MLPRegressor models
model_age_to_years_of_exp = MLPRegressor(hidden_layer_sizes=(50, 50), max_iter=1000)

# training the MLPRegressor models
model_age_to_years_of_exp.fit(X_train_age_to_yearsofexp, y_train_year_of_experience)

# Make predictions
predictions_age_to_year_of_experience = model_age_to_years_of_exp.predict(X_test_age_to_yearsofexp)

# Print the predictions
# print("Predicte Years of Experience based on Age:", predictions_age_to_year_of_experience)

predictions_df = pd.DataFrame({
    'Predicted Years of Experience based on Age ': predictions_age_to_year_of_experience
})

print(predictions_df)
print("Mean Score Predictions: ", model_age_to_years_of_exp.score(X_test_age_to_yearsofexp, y_test_year_of_experience))
print("Mean Absolute Error:",mean_absolute_error(y_test_year_of_experience, predictions_age_to_year_of_experience))


C:\Users\HP\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:1617: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


     Predicted Years of Experience based on Age 
0                                      10.107935
1                                       6.025400
2                                       3.815729
3                                      14.190470
4                                       9.291428
..                                           ...
353                                    14.190470
354                                     7.658414
355                                     6.025400
356                                    17.456499
357                                    12.557456

[358 rows x 1 columns]
Mean Score Predictions:  0.8792670043146954
Mean Absolute Error: 1.7569969984272096


### Given that the model is unable to generate predictions for categorical variables which are "Job Title, Education Level and Gender" represented as strings, I needed to apply one-hot encoding to transform the textual data into numerical format.

### Predictions of Salary from Job Title

In [4]:
# Reading the data
salary_df = pd.read_csv("Clean Database.csv")

# features and target variables
X_job_title_salary = salary_df[['Job Title']]
y_salary_from_job_title = salary_df[['Salary']]


# Convert the job titles to numerical using one-hot encoding
encoder = OneHotEncoder()
X_encoded_job_title_salary = encoder.fit_transform(X_job_title_salary).toarray()

# Split the data into training and testing sets
X_train_job_title_salary, X_test_job_title_salary, y_train_salary_from_job_title, y_test_salary_from_job_title = train_test_split(X_encoded_job_title_salary, y_salary_from_job_title, test_size=0.2)


# Initializing the MLPRegressor models
model_job_title_salary = MLPRegressor(hidden_layer_sizes=(50, 50), max_iter=1000)

# training the MLPRegressor models
model_job_title_salary.fit(X_train_job_title_salary, y_train_salary_from_job_title)

# Make predictions
predictions_salary_from_job_title = model_job_title_salary.predict(X_test_job_title_salary)

# Print the predictions
# print("Predictions on Salary from Job Title:", predictions_salary_from_job_title)

predictions_df = pd.DataFrame({
    'Predicted Salary from Job Titlee ': predictions_salary_from_job_title
})

print(predictions_df)
print("Mean Score Predictions: ", model_job_title_salary.score(X_test_job_title_salary, y_test_salary_from_job_title))


C:\Users\HP\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:1617: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


     Predicted Salary from Job Titlee 
0                        153415.246425
1                        128445.546272
2                        165618.272379
3                        165618.272379
4                         99113.578448
..                                 ...
353                       45014.807063
354                      163173.620303
355                      126741.320498
356                      153415.246425
357                       99113.578448

[358 rows x 1 columns]
Mean Score Predictions:  0.5900950967691527


C:\Users\HP\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:684: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(


### Predictions of Years of Exprience from Job Title

In [5]:
# Reading the data
salary_df = pd.read_csv("Clean Database.csv")

# features and target variables
X_job_title_yearsofexp = salary_df[['Job Title']]
y__yearsofexp_from_job_title = salary_df['Years of Experience']


# Convert the job titles to numerical using one-hot encoding
encoder = OneHotEncoder()
X_encoded_job_title_yearsofexp = encoder.fit_transform(X_job_title_yearsofexp).toarray()

# Split the data into training and testing sets
X_train_job_title_yearsofexp, X_test_job_title_yearsofexp, y_train__yearsofexp_from_job_title, y_test__yearsofexp_from_job_title = train_test_split(X_encoded_job_title_yearsofexp, y__yearsofexp_from_job_title, test_size=0.2)


# Initializing the MLPRegressor models
model_job_title_years_of_exp = MLPRegressor(hidden_layer_sizes=(50, 50), max_iter=1000)

# training the MLPRegressor models
model_job_title_years_of_exp.fit(X_train_job_title_yearsofexp, y_train__yearsofexp_from_job_title)

# Make predictions
predictions_Yearsofexp_from_job_title = model_job_title_years_of_exp.predict(X_test_job_title_yearsofexp)

# Print the predictions
# print("Predictions on Years of Exprience from Job Title:", predictions_Yearsofexp_from_job_title)

predictions_df = pd.DataFrame({
    'Predicted Years of Experience based on Job Title ': predictions_Yearsofexp_from_job_title
})

print(predictions_df)
print("Mean Score Predictions: ", model_job_title_salary.score(X_test_job_title_salary, y_test_salary_from_job_title))



     Predicted Years of Experience based on Job Title 
0                                            12.965759
1                                             3.576033
2                                            14.220468
3                                             6.385562
4                                             8.848673
..                                                 ...
353                                           6.939571
354                                           4.388005
355                                           3.035818
356                                           9.375227
357                                          14.220468

[358 rows x 1 columns]
Mean Score Predictions:  0.5900950967691527


### Predictions of Years of Exprience from Education Level

In [6]:
# Reading the data
salary_df = pd.read_csv("Clean Database.csv")

# features and target variables
X_educ_lvl_yearsofexp = salary_df[['Education Level']]
y__yearsofexp_from_educ_lvl = salary_df[['Years of Experience']]


# Convert the Educationg Level to numerical using one-hot encoding
encoder = OneHotEncoder()
X_encoded_educ_lvl_yearsofexp = encoder.fit_transform(X_educ_lvl_yearsofexp).toarray()

# Split the data into training and testing sets
X_train_educ_lvl_yearsofexp, X_test_educ_lvl_yearsofexp, y_train__yearsofexp_from_educ_lvl, y_test__yearsofexp_from_educ_lvl = train_test_split(X_encoded_educ_lvl_yearsofexp, y__yearsofexp_from_educ_lvl, test_size=0.2)


# Initializing the MLPRegressor models
model_education_lvl_years_of_exp = MLPRegressor(hidden_layer_sizes=(50, 50), max_iter=1000)

# training the MLPRegressor models
model_education_lvl_years_of_exp.fit(X_train_educ_lvl_yearsofexp, y_train__yearsofexp_from_educ_lvl)

# Make predictions
predictions_Yearsofexp_from_education_level = model_education_lvl_years_of_exp.predict(X_test_educ_lvl_yearsofexp)

# Print the predictions
# print("Predictions on Years of Exprience from Education level:", predictions_Yearsofexp_from_education_level)

predictions_df = pd.DataFrame({
    'Predicted Years of Experience based on Education level ': predictions_Yearsofexp_from_education_level
})

print(predictions_df)
print("Mean Score Predictions: ",model_education_lvl_years_of_exp.score(X_test_educ_lvl_yearsofexp, y_test__yearsofexp_from_educ_lvl))


C:\Users\HP\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:1617: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


     Predicted Years of Experience based on Education level 
0                                             6.542018      
1                                            15.031560      
2                                            10.328090      
3                                             6.542018      
4                                            10.328090      
..                                                 ...      
353                                          15.031560      
354                                          10.328090      
355                                           6.542018      
356                                          10.328090      
357                                          10.328090      

[358 rows x 1 columns]
Mean Score Predictions:  0.21901769712856878


### Predictions of Salary from Education Level

In [7]:
# Reading the data
salary_df = pd.read_csv("Clean Database.csv")

# features and target variables
X_educ_lvl_salary = salary_df[['Education Level']]
y__salary_from_educ_lvl = salary_df[['Salary']]


# Convert the Educationg Level to numerical using one-hot encoding
encoder = OneHotEncoder()
X_encoded_educ_lvl_salary = encoder.fit_transform(X_educ_lvl_salary).toarray()

# Split the data into training and testing sets
X_train_educ_lvl_salary, X_test_educ_lvl_salary, y_train__salary_from_educ_lvl, y_test__salary_from_educ_lvl = train_test_split(X_encoded_educ_lvl_salary, y__salary_from_educ_lvl, test_size=0.2)


# Initializing the MLPRegressor models
model_education_lvl_Salary = MLPRegressor(hidden_layer_sizes=(50, 50), max_iter=1000)

# training the MLPRegressor models
model_education_lvl_Salary.fit(X_train_educ_lvl_salary, y_train__salary_from_educ_lvl)

# Make predictions
predictions_Salary_from_education_level = model_education_lvl_Salary.predict(X_test_educ_lvl_salary)

# Print the predictions
# print("Predictions on Salary from Education level:", predictions_Salary_from_education_level)

predictions_df = pd.DataFrame({
    'Predicted Salary based on Education level ': predictions_Salary_from_education_level
})

print(predictions_df)
print("Mean Score Predictions: ",model_education_lvl_Salary.score(X_test_educ_lvl_salary, y_test__salary_from_educ_lvl))

C:\Users\HP\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:1617: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


     Predicted Salary based on Education level 
0                                  90900.628644
1                                  90900.628644
2                                 158643.464399
3                                  90900.628644
4                                  90900.628644
..                                          ...
353                               128426.344221
354                               128426.344221
355                                90900.628644
356                               128426.344221
357                                90900.628644

[358 rows x 1 columns]
Mean Score Predictions:  0.4007797511738548


# Decision Tree Model

### i picked the decision tree model because they're easy to understand, just like making decisions by asking simple questions, and they can work well when you have a bunch of features that help you classify things.

### the Mean Absolute Error (MAE) is a valuable metric for evaluating the performance of regression models,which gives you insights into the average discrepancy between predicted and actual values. I could not calculate the Mean absolute Error for the categorical variables becasue this model does not permit it

### Under this model using the database given, The code below is for predicting Salary from Age, As shown below the MAE shows that the prediction of Salary is 26,245k off and the MSP is 57% so far Neural Networks is better than Decision Tree Model in predicting

In [8]:
# Extracting the columns
X_age_salary = salary_df[['Age']]
y_salary = salary_df['Salary']

# Split the data into training and testing sets
X_age_salary_train, X_age_salary_test, y_salary_train, y_salary_test = train_test_split(X_age_salary, y_salary, test_size=0.2, random_state=42)

# Training the Decision Tree models 
age_salary_tree = DecisionTreeRegressor(random_state=42)
age_salary_tree.fit(X_age_salary_train, y_salary_train)

# Make predictions
predictions_age_salary = age_salary_tree.predict(X_age_salary_test)
print("Mean Absolute Error:",mean_absolute_error(y_salary_test, predictions_age_salary))
print("Mean Score Predictions: ",age_salary_tree.score(X_age_salary_test, y_salary_test))

#Tabularizing the predictions
predictions_df = pd.DataFrame({
    'Predicted Years of Experience based on Age ': predictions_age_salary
})

print(predictions_df)


Mean Absolute Error: 26245.72628494613
Mean Score Predictions:  0.5376578618563552
     Predicted Years of Experience based on Age 
0                                   78795.954545
1                                  124551.210526
2                                  145415.604651
3                                   71376.329114
4                                  194381.750000
..                                           ...
353                                110149.831325
354                                 78795.954545
355                                 75792.405405
356                                110149.831325
357                                155008.311111

[358 rows x 1 columns]


### Predicted Years of Experience based on Age

In [9]:
# Extracting the columns
X_age_salary = salary_df[['Age']]
y_salary = salary_df['Years of Experience']

# Split the data into training and testing sets
X_age_salary_train, X_age_salary_test, y_salary_train, y_salary_test = train_test_split(X_age_salary, y_salary, test_size=0.2, random_state=42)

# Training the Decision Tree models 
age_salary_tree = DecisionTreeRegressor(random_state=42)
age_salary_tree.fit(X_age_salary_train, y_salary_train)

# Make predictions
predictions_age_salary = age_salary_tree.predict(X_age_salary_test)
print("Mean Absolute Error:",mean_absolute_error(y_salary_test, predictions_age_salary))
print("Mean Score Predictions: ",age_salary_tree.score(X_age_salary_test, y_salary_test))

#Tabularizing the predictions
predictions_df = pd.DataFrame({
    'Predicted Years of Experience based on Age ': predictions_age_salary
})

print(predictions_df)


Mean Absolute Error: 1.8152152876148295
Mean Score Predictions:  0.8805843409608359
     Predicted Years of Experience based on Age 
0                                       3.926136
1                                      10.026316
2                                      14.325581
3                                       2.841772
4                                      26.250000
..                                           ...
353                                     7.493976
354                                     3.926136
355                                     3.412162
356                                     7.493976
357                                    16.133333

[358 rows x 1 columns]


### Predictions Years of Experience based on Salary, , Based on this predictions the MAE shows that the prediction of Years of Experience to Salary is off by 3 and the MSP is 69%

In [10]:
# Extracting the columns
X_salary_yearsofexp = salary_df[['Salary']]
y_yearsofexp = salary_df['Years of Experience']

# Split the data into training and testing sets
X_salary_yearsofexp_train, X_salary_yearsofexp_test, y_yearsofexp_train, y_yearsofexp_test = train_test_split(X_salary_yearsofexp, y_yearsofexp, test_size=0.2, random_state=42)

# Training the Decision Tree models 
Salary_Years_of_exp_tree = DecisionTreeRegressor(random_state=42)
Salary_Years_of_exp_tree.fit(X_salary_yearsofexp_train, y_yearsofexp_train)

# Make predictions
predictions_Years_of_exp = Salary_Years_of_exp_tree.predict(X_salary_yearsofexp_test)
print("Mean Absolute Error:", mean_absolute_error(y_yearsofexp_test, predictions_Years_of_exp))
print("Mean Score Predictions: ",Salary_Years_of_exp_tree.score(X_salary_yearsofexp_test, y_yearsofexp_test))
                
#Tabularizing the predictions
predictions_df = pd.DataFrame({
    'Predicted Years of Experience based on Salary ': predictions_Years_of_exp
})

print(predictions_df)

Mean Absolute Error: 2.7140857801277622
Mean Score Predictions:  0.6905740395258179
     Predicted Years of Experience based on Salary 
0                                          1.836957
1                                          6.000000
2                                          9.560976
3                                          2.686047
4                                         15.760000
..                                              ...
353                                        9.318182
354                                       13.000000
355                                        2.000000
356                                       12.000000
357                                       21.000000

[358 rows x 1 columns]


### Predictions on Gender based on Job Title

In [11]:
# Extracting the columns
X_job_title = salary_df[['Job Title']]
y_gender = salary_df['Gender']

# Apply one-hot encoding to job titles
X_encoded_job_title = pd.get_dummies(X_job_title, drop_first=True)

# Split the data into training and testing sets
X_job_title_train, X_job_title_test, y_gender_train, y_gender_test = train_test_split(X_encoded_job_title, y_gender, test_size=0.2, random_state=42)

# Training the Decision Tree models 
gender_tree = DecisionTreeClassifier(random_state=42)
gender_tree.fit(X_job_title_train, y_gender_train)

# Make predictions
predictions_gender = gender_tree.predict(X_job_title_test)
# print("Mean Absolute Error:", mean_absolute_error(y_gender_test, predictions_gender))

#Tabularizing the predictions
predictions_df = pd.DataFrame({
    'Predicted Gender based on Job Title ': predictions_gender
})

print(predictions_df)


    Predicted Gender based on Job Title 
0                                   Male
1                                   Male
2                                 Female
3                                   Male
4                                   Male
..                                   ...
353                                 Male
354                                 Male
355                                 Male
356                                 Male
357                                 Male

[358 rows x 1 columns]


### Predictions on Education Level based on Job Title

In [12]:
# Extracting the columns
X_job_title = salary_df[['Job Title']]
y_edu_lvl = salary_df['Education Level']

# Apply one-hot encoding to job titles
X_encoded_job_edu_lvl = pd.get_dummies(X_job_title, drop_first=True)
y_encoded_edu_lvl = pd.get_dummies(y_edu_lvl, drop_first=True)


# Split the data into training and testing sets
X_job_title_train, X_job_title_test, y_edu_lvl_train, y_edu_lvl_test = train_test_split(X_encoded_job_edu_lvl, y_edu_lvl, test_size=0.2, random_state=42)

# Training the Decision Tree models 
Educationlvl_tree = DecisionTreeClassifier(random_state=42)
Educationlvl_tree.fit(X_job_title_train, y_edu_lvl_train)

# Make predictions
predictions_Education_Level = Educationlvl_tree.predict(X_job_title_test)
# print("Mean Absolute Error:", mean_absolute_error(y_encoded_edu_lvl, predictions_Education_Level))

#Tabularizing the predictions
predictions_df = pd.DataFrame({
    'Predicted  Education Level based on Job Title ': predictions_Education_Level
})

print(predictions_df)

    Predicted  Education Level based on Job Title 
0                               Bachelor's Degree 
1                               Bachelor's Degree 
2                               Bachelor's Degree 
3                               Bachelor's Degree 
4                                              PhD
..                                             ...
353                                            PhD
354                             Bachelor's Degree 
355                             Bachelor's Degree 
356                             Bachelor's Degree 
357                               Master's Degree 

[358 rows x 1 columns]


# Linear Regression

###  i picked linear regression model because it helps in making predictions about one thing (like your earnings i.e salary) based on another thing (like Years of experience or age), by drawing a straight line that helps you see the connection

### Predicted Years of Experience based on Age

In [13]:
from sklearn.metrics import mean_absolute_error

# Extracting the features and target variables
X_age_yearsofexp = salary_df[['Age']]
y_yearsofexp = salary_df['Years of Experience']

# Initializing the Linear Regression models
model_age_yearsofexp = LinearRegression()

# Spliting the data into training and testing sets
X_train_age_yearsofexp, X_test_age_yearsofexp, y_train_yearsofexp, y_test_yearsofexp = train_test_split(X_age_yearsofexp, y_yearsofexp, random_state= 0)

# Training the models
model_age_yearsofexp.fit(X_train_age_yearsofexp, y_train_yearsofexp)

# Make predictions
predictions_Years_of_Exp = model_age_yearsofexp.predict(X_train_age_yearsofexp)

print("Mean Score Predictions: ", model_age_yearsofexp.score(X_test_age_yearsofexp, y_test_yearsofexp))
# print("Mean Absolute Error:", mean_absolute_error(y_test_yearsofexp, predictions_Years_of_Exp))


predictions_df = pd.DataFrame({
    'Predicted Years of Experience based on Age ': predictions_Years_of_Exp
})

print(predictions_df)


Mean Score Predictions:  0.8779470641817055
      Predicted Years of Experience based on Age 
0                                        3.590027
1                                        4.378663
2                                       16.996833
3                                       20.940011
4                                       12.265019
...                                           ...
1335                                     7.533206
1336                                    11.476384
1337                                    20.151375
1338                                     5.167299
1339                                     5.955934

[1340 rows x 1 columns]


### Predicted  Salary based on Years of Experience

In [14]:
# Extracting the features and target variables
X_yearsofexp_salary = salary_df[['Years of Experience']]
y__salary = salary_df['Salary']

# Initializing the Linear Regression models
model_yearsofexp_salary = LinearRegression()

# Spliting the data into training and testing sets
X_train_yearsofexp_salary, X_test_yearsofexp_salary, y_train__salary, y_test__salary = train_test_split(X_yearsofexp_salary, y__salary, test_size=0.2)

# Training the models
model_yearsofexp_salary.fit(X_train_yearsofexp_salary, y_train__salary)

# Make predictions
prediction_salary = model_yearsofexp_salary.predict(X_train_yearsofexp_salary)

# print("Mean Score Predictions: ", model_yearsofexp_salary.score(X_yearsofexp_salary, y_test__salary))
# print("Mean Absolute Error:", mean_absolute_error(y_test__salary, prediction_salary))

predictions_df = pd.DataFrame({
    'Predicted  Salary based on Years of Experience ': predictions_salary
})

print(predictions_df)

     Predicted  Salary based on Years of Experience 
0                                      145386.306207
1                                      115225.477232
2                                      142035.102987
3                                      135332.696549
4                                      135332.696549
..                                               ...
353                                     85064.648257
354                                     91767.054696
355                                     91767.054696
356                                     98469.461134
357                                     98469.461134

[358 rows x 1 columns]


# Comparison

## Using these three models and considering the Mean Score Predictions and Mean Absolute Error (MAE), specifically for "Years of Experience Based on Age," it has come to my attention that:

### The Mean Score predictions on Years of Experience based on Age are:
### Decision Tree Model : 88%
### Neural Network: 87%
### Linear Regression : 88%

### The Mean Absolute Error (MAE) on Years of Experience based on Age are:

### Decision Tree Model : 2

### Neural Network: 2

### Linear Regression : 3 



### In summary, given the comparable mean score predictions and MAE values across all three models, it's challenging to determine which one is better among them